src/pykeen/nn/emb.py

```diff
@@ -1068,3 +1070,157 @@ def forward(
+        size=(triples_factory.num_entities, num_tokens),
+        dtype=torch.long,
+        fill_value=-1,
+    )
+    for e, rs in e2r.items():
+        rs = torch.as_tensor(data=list(rs), dtype=torch.long)
+        rs = _sample(rs=rs, k=num_tokens)
+        assignment[e, : len(rs)] = rs
+
+    return assignment
+
+
+def resolve_aggregation(
+    aggregation: Union[None, str, Callable[[torch.FloatTensor, int], torch.FloatTensor]],
+) -> Callable[[torch.FloatTensor, int], torch.FloatTensor]:
+    """Resolve the aggregation function."""
+    if aggregation is None:
+        return torch.mean
+
+    if isinstance(aggregation, str):
+        return getattr(torch, aggregation)
+
+    return aggregation
+
+
+class NodePieceRepresentation(RepresentationModule):
+    r"""
+    Basic implementation of node piece decomposition [galkin2021]_.
+
+    .. math ::
+        x_e = agg(\{T[t] \mid t \in tokens(e) \})
+
+    where $T$ are token representations, $tokens$ selects a fixed number of $k$ tokens for each entity, and $agg$ is
+    an aggregation function, which aggregates the individual token representations to a single entity representation.
+
+    .. note ::
+        This implementation currently only supports representation of entities by bag-of-relations.
+    """
+
+    #: the token representations
+    tokens: RepresentationModule
+
+    #: the entity-to-token mapping
+    assignment: torch.LongTensor
+
+    def __init__(
+        self,
+        *,
+        triples_factory: CoreTriplesFactory,
+        token_representation: Union[EmbeddingSpecification, RepresentationModule],
+        aggregation: Union[None, str, Callable[[torch.FloatTensor, int], torch.FloatTensor]] = None,
+        num_tokens: int = 2,
+        shape: Optional[Sequence[int]] = None,
+    ):
+        """
+        Initialize the representation.
+
+        :param triples_factory:
+            the triples factory
+        :param token_representation:
+            the token representation specification, or pre-instantiated representation module. For the latter, the
+            number of representations must be $2 * num_relations + 1$.
+        :param aggregation:
+            aggregation of multiple token representations to a single entity representation. By default,
+            this uses :func:`torch.mean`. It could also use other aggregations like :func:`torch.sum`,
+            :func:`torch.max`, or even trainable aggregations e.g., ``MLP(mean(MLP(tokens)))``
+            (cf. DeepSets from [zaheer2017]_) if given value ``"mlp"``.
+
+            We could also have aggregations which result in differently shapes output, e.g. a concatenation of all
+            token embeddings resulting in shape ``(num_tokens * d,)``. In this case, `shape` must be provided.
+
+            The aggregation takes two arguments: the (batched) tensor of token representations, in shape
+            ``(*, num_tokens, *dt)``, and the index along which to aggregate.
+        :param shape:
```

Comment:
Do we have a compelling use case for this? Otherwise I think it's quite a confusing setting

Reply:
citing from a few lines above 😉 
https://github.com/pykeen/pykeen/blob/4ee90a30ddad81051122923a6c5ee3599b84e894/src/pykeen/models/unimodal/node_piece.py#L113-L114

Reply:
lol RTFD, charlie 😆